core/domain/summary_services.py

```diff
@@ -144,6 +144,7 @@ def get_learner_collection_dict_by_id(
         next_exploration_ids = collection.init_exploration_ids
         completed_exp_ids = []
 
+    collection.sort_nodes()
```

Comment:
Better to do this in to_dict() perhaps? And also in validate() (before saving)?

That way, we guarantee that any dict representation of the collection is correctly ordered, and also that subsequent saves of the collection are ordered correctly.

Reply:
Actually, scrap that comment. Per offline discussion, here's what we came to:
- One approach is to think of the nodes in the backend as an unordered dict, and then do transient ordering of them in the frontend before showing it to the user. I.e. the client gets the unordered nodes from the backend, and sorts them before presenting anything. However, this approach has weaknesses: topological sorts aren't unique (so the creator may see something different from the learner), and in this approach the creator doesn't have much control over how the nodes actually get sorted.
- So, instead, we propose the following approach: every time a collection is saved to the datastore, we should ensure that its nodes are sorted properly. This has three implications:
  - In the learner view, we just display collection nodes in the order they are in the backend; no additional processing is needed here.
  - In the creator view, if the nodes are out of order, we show a validation warning to the creator saying "students can't do exp A at this point because they won't have the skills yet".
  - In the backend, before we save a new version of a collection, we check that the lessons can be completed in order (i.e. all prereq skills for that collection nodes must have already been acquired in previous lessons). If this constraint is not met, we reject the save. A similar check should be done in the frontend for the collection editor prior to sending the nodes to the backend.